tests/MongoDB.Bson.Tests/IO/JsonReaderTests.cs

```diff
@@ -116,18 +121,114 @@ public void TestArrayTwoElements()
+        [InlineData("{ $binary : { subType : \"0\", base64 : \"AQ==\"")]
+        [InlineData("{ $binary : { subType : \"0\", base64 : \"AQ==\" }")]
+        [InlineData("BinData")]
+        [InlineData("BinData(")]
+        [InlineData("BinData(0")]
+        [InlineData("BinData(0,")]
+        [InlineData("BinData(0, \"AQ==\"")]
+        [InlineData("HexData")]
+        [InlineData("HexData(")]
+        [InlineData("HexData(0")]
+        [InlineData("HexData(0,")]
+        [InlineData("HexData(0, \"01\"")]
+        public void TestBinaryDataEndOfFile(string json)
+        {
+            using (var reader = new JsonReader(json))
+            {
+                var exception = Record.Exception(() => reader.ReadBinaryData());
+
+                exception.Should().BeOfType<FormatException>();
+            }
+        }
+
+        [Theory]
+        [InlineData("{ $binary [")]
+        [InlineData("{ $binary : [")]
+        [InlineData("{ $binary : \"AQ==\" [")]
+        [InlineData("{ $binary : \"AQ==\", [")]
+        [InlineData("{ $binary : \"AQ==\", $type [")]
+        [InlineData("{ $binary : \"AQ==\", $type : [")]
+        [InlineData("{ $binary : \"AQ==\", $type : 0 [")]
+        [InlineData("{ $binary : { [")]
+        [InlineData("{ $binary : { base64 [")]
+        [InlineData("{ $binary : { base64 : [")]
+        [InlineData("{ $binary : { base64 : \"AQ==\" [")]
+        [InlineData("{ $binary : { base64 : \"AQ==\", [")]
+        [InlineData("{ $binary : { base64 : \"AQ==\", subType [")]
+        [InlineData("{ $binary : { base64 : \"AQ==\", subType : [")]
+        [InlineData("{ $binary : { base64 : \"AQ==\", subType : \"0\" [")]
+        [InlineData("{ $binary : { base64 : \"AQ==\", subType : \"0\" } [")]
+        [InlineData("{ $binary : { subType [")]
+        [InlineData("{ $binary : { subType : [")]
+        [InlineData("{ $binary : { subType : \"0\" [")]
+        [InlineData("{ $binary : { subType : \"0\", [")]
+        [InlineData("{ $binary : { subType : \"0\", base64 [")]
+        [InlineData("{ $binary : { subType : \"0\", base64 : [")]
+        [InlineData("{ $binary : { subType : \"0\", base64 : \"AQ==\" [")]
+        [InlineData("{ $binary : { subType : \"0\", base64 : \"AQ==\" } [")]
+        [InlineData("BinData [")]
+        [InlineData("BinData( [")]
+        [InlineData("BinData(0 [")]
+        [InlineData("BinData(0, [")]
+        [InlineData("BinData(0, \"AQ==\" [")]
+        [InlineData("HexData [")]
+        [InlineData("HexData( [")]
+        [InlineData("HexData(0 [")]
+        [InlineData("HexData(0, [")]
+        [InlineData("HexData(0, \"01\" [")]
+        public void TestBinaryDataInvalidToken(string json)
+        {
+            using (var reader = new JsonReader(json))
+            {
+                var exception = Record.Exception(() => reader.ReadBinaryData());
+
+                exception.Should().BeOfType<FormatException>();
+            }
+        }
```

Comment:
Note that the bodies of the previous two tests are identical. Even so, I think it's better to have two separate tests since hitting end of file is sort of different than encountering an invalid token.